lib/core/src/Cardano/Wallet/Primitive/Scripts.hs

```diff
@@ -0,0 +1,118 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of wallet's verification key discovery
+-- in a script.
+
+module Cardano.Wallet.Primitive.Scripts
+    ( isShared
+    , retrieveAllVerKeyHashes
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( KeyHash (..), Script (..), ScriptHash (..), toScriptHash )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationType (..)
+    , Index (..)
+    , Role (..)
+    , SoftDerivation (..)
+    )
+import Cardano.Wallet.Primitive.AddressDerivation.Shelley
+    ( ShelleyKey (..) )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( SeqState (..)
+    , VerificationKeyPool (..)
+    , defaultAddressPoolGap
+    , getAddressPoolGap
+    , mkUnboundedAddressPoolGap
+    , mkVerificationKeyPool
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( AddressState (..) )
+import Control.Monad
+    ( foldM )
+import Data.Functor.Identity
+    ( Identity (..) )
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+instance Ord ScriptHash where
+    compare (ScriptHash sh1) (ScriptHash sh2) = compare sh1 sh2
+
+isShared
+    :: (k ~ ShelleyKey, SoftDerivation k)
+    => Script
+    -> SeqState n k
+    -> ([k 'ScriptK XPub], SeqState n k)
+isShared script (SeqState !s1 !s2 !ixs !rpk !prefix !scripts !s3) =
+    let verKeysInScript = retrieveAllVerKeyHashes script
+        (VerificationKeyPool accXPub currentGap verKeyMap) = s3
```

Comment:
I don't think this should be exposed. It'd be cleaner to define functions that work with `VerificationKeyPool` next to the definition of `VerificationKeyPool` in the same way we do not know how a `Map` or `HashMap` is implemented. Instead, we rely on the various functions exposed which serves as a contract. The same goes for `AddressPool`. Any manipulation of an `AddressPool` is done through specific functions:

- mkAddressPool
- lookupAddress
- shrinkPool

This way, nothing leaks out of the AddressPool module and it makes it: (a) easier to test, (b) easier to compose with other things.

Reply:
added getters https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/e3e413b5003318e5827b1d829d0470fd87b54f90